src/realm/geospatial.hpp

```diff
@@ -140,13 +187,103 @@ class Geospatial {
 
 private:
     Type m_type;
-    std::optional<std::string> m_invalid_type;
-
     std::vector<GeoPoint> m_points;
     std::optional<double> m_radius_radians;
 
     mutable std::shared_ptr<S2Region> m_region;
     S2Region& get_region() const;
+
+    friend class GeospatialRef;
+};
+
+template <>
+inline GeoCenterSphere Geospatial::get<GeoCenterSphere>() const noexcept
+{
+    REALM_ASSERT_EX(m_type == Type::CenterSphere, get_type_string());
+    REALM_ASSERT(m_radius_radians);
+    REALM_ASSERT(m_points.size() >= 1);
+    return GeoCenterSphere{*m_radius_radians, m_points[0]};
+}
+
+template <>
+inline GeoPolygon Geospatial::get<GeoPolygon>() const noexcept
+{
+    REALM_ASSERT_EX(m_type == Type::Polygon, get_type_string());
+    REALM_ASSERT(m_points.size() >= 1);
+    return GeoPolygon(m_points);
+}
+
+class GeospatialRef {
+public:
+    GeospatialRef(const GeoPoint* data, const Geospatial& geo)
+        : m_data(data)
+        , m_sphere_radius(geo.m_radius_radians ? *geo.m_radius_radians : 0)
+        , m_size(unsigned(geo.m_points.size()))
+        , m_type(geo.m_type)
+    {
+    }
+    GeospatialRef(const GeoPoint* data, size_t size, Geospatial::Type type, std::optional<double> sphere_radius)
+        : m_data(data)
+        , m_sphere_radius(sphere_radius ? *sphere_radius : 0)
+        , m_size(unsigned(size))
+        , m_type(type)
+    {
+    }
+    Geospatial get() const
+    {
+        REALM_ASSERT(m_data);
+        switch (m_type) {
+            case Geospatial::Type::Invalid:
+                return Geospatial(); // FIXME: error handling, should this throw instead?
```

Comment:
REALM_UNREACHABLE? We shouldn't hit this path from what i can tell, apart from constructing GeospatialRef from invalid Geospatial, I don't think there is the usecase like this. So the constructor should throw instead InvalidArgument or something.

Reply:
I think this is a reachable and valid case if a Geospatial query argument is invalid. I added a test for it.